src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/services/EventMessageReceiverService.java

```diff
@@ -0,0 +1,134 @@
+import static net.logstash.logback.encoder.org.apache.commons.lang3.StringUtils.isNotBlank;
+
+@Slf4j
+@Service
+@Transactional
+public class EventMessageReceiverService {
+    public static final String MESSAGE_PROPERTIES = "MessageProperties";
+
+    private final ObjectMapper objectMapper;
+    private final CaseEventMessageRepository repository;
+    private final CaseEventMessageMapper mapper;
+
+    public EventMessageReceiverService(ObjectMapper objectMapper,
+                                       CaseEventMessageRepository repository,
+                                       CaseEventMessageMapper caseEventMessageMapper) {
+        this.objectMapper = objectMapper;
+        this.repository = repository;
+        this.mapper = caseEventMessageMapper;
+    }
+
+    public CaseEventMessage handleDlqMessage(String messageId, String message) {
+        log.info("Received DLQ message with id '{}'", messageId);
+        return handleMessage(messageId, message, true);
+    }
+
+    public CaseEventMessage handleAsbMessage(String messageId, String message) {
+        log.info("Received ASB message with id '{}'", messageId);
+        return handleMessage(messageId, message, false);
+    }
+
+    public CaseEventMessage getMessage(String messageId) {
+        CaseEventMessageEntity messageEntity = repository.findByMessageId(messageId).stream().findFirst()
+            .orElseThrow(() -> new CaseEventMessageNotFoundException(
+                format("Could not find a message with message_id: %s", messageId)));
+        return mapper.mapToCaseEventMessage(messageEntity);
+    }
+
+    private CaseEventMessage handleMessage(String messageId, String message, boolean fromDlq) {
+
+        try {
+            EventInformation eventInformation = objectMapper.readValue(message, EventInformation.class);
+
+            boolean isValid = validate(messageId, eventInformation);
+            CaseEventMessageEntity messageEntity;
+            if (isValid) {
+                messageEntity = build(messageId, message, fromDlq, eventInformation, MessageState.NEW);
+            } else {
+                messageEntity = build(messageId, message, fromDlq, MessageState.UNPROCESSABLE);
+            }
+            repository.save(messageEntity);
+
+            log.info("Message with id '{}' successfully stored into the DB", messageId);
+
+            return mapper.mapToCaseEventMessage(messageEntity);
+        } catch (JsonProcessingException e) {
+            log.error("Could not parse the message with id '{}'", messageId);
+
+            CaseEventMessageEntity messageEntity = build(messageId, message, fromDlq, MessageState.UNPROCESSABLE);
+            repository.save(messageEntity);
+
+            return mapper.mapToCaseEventMessage(messageEntity);
+        } catch (DataIntegrityViolationException e) {
+            throw new CaseEventMessageDuplicateMessageIdException(
+                format("Trying to save a message with a duplicate messageId: %s", messageId), e);
```

Comment:
This is probably not a requirement of this story so don't need to address but to clarify.  
We don't want to throw an error if we get a duplicate message.  We want to simply increment the `delivery_count` and successfully consume the message received from ASB.

Reply:
This will mean that the `insert` statement will have to be a bit more specific.  E.g. have an `on conflict (message_id) do update . . .` clause 

see: https://www.postgresql.org/docs/11/sql-insert.html